orchestrators/kubernetes.go

```diff
@@ -0,0 +1,299 @@
+func NewKubernetesOrchestrator(c *handler.Conplicity) (o *KubernetesOrchestrator) {
+	var err error
+	o = &KubernetesOrchestrator{
+		Handler: c,
+	}
+
+	config, err := o.getConfig()
+	if err != nil {
+		log.Fatalf("failed to retrieve Kubernetes config: %s", err)
+	}
+
+	o.Client, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Errorf("failed to create a Kubernetes client: %v", err)
+	}
+	return
+}
+
+// GetName returns the orchestrator name
+func (*KubernetesOrchestrator) GetName() string {
+	return "Kubernetes"
+}
+
+// GetHandler returns the Orchestrator's handler
+func (o *KubernetesOrchestrator) GetHandler() *handler.Conplicity {
+	return o.Handler
+}
+
+// GetVolumes returns the Kubernetes persistent volume claims, inspected and filtered
+func (o *KubernetesOrchestrator) GetVolumes() (volumes []*volume.Volume, err error) {
+	c := o.Handler
+
+	pvcs, err := o.Client.CoreV1().PersistentVolumeClaims(o.Handler.Config.Kubernetes.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		log.Errorf("failed to retrieve the list of PVCs: %v", err)
+	}
+
+	containers, err := o.GetMountedVolumes()
+	mountedVolumes := make(map[string]string)
+	for _, container := range containers {
+		for volName, volMountpath := range container.Volumes {
+			mountedVolumes[volName] = volMountpath
+		}
+	}
+	var mountpoint string
+	for _, pvc := range pvcs.Items {
+		if value, ok := mountedVolumes[pvc.Name]; ok {
+			mountpoint = value
+		} else {
+			mountpoint = "/data"
+		}
+		nv := &volume.Volume{
+			Config:     &volume.Config{},
+			Mountpoint: mountpoint,
+			Name:       pvc.Name,
+		}
+
+		v := volume.NewVolume(nv, c.Config, c.Hostname)
+		if b, r, s := o.blacklistedVolume(v); b {
+			log.WithFields(log.Fields{
+				"volume": pvc.Name,
+				"reason": r,
+				"source": s,
+			}).Info("Ignoring volume")
+			continue
+		}
+		volumes = append(volumes, v)
+		log.Infof("%+v", v)
+	}
+	return
+}
+
+// LaunchContainer starts a container using the Kubernetes orchestrator
+func (o *KubernetesOrchestrator) LaunchContainer(image string, env map[string]string, cmd []string, volumes []*volume.Volume) (state int, stdout string, err error) {
+
+	var envVars []apiv1.EnvVar
+	for envName, envValue := range env {
+		ev := apiv1.EnvVar{
+			Name:  envName,
+			Value: envValue,
+		}
+		envVars = append(envVars, ev)
+	}
+
+	kvs := []apiv1.Volume{}
+	kvms := []apiv1.VolumeMount{}
+	for _, v := range volumes {
+		kv := apiv1.Volume{
+			Name: v.Name,
+			VolumeSource: apiv1.VolumeSource{
+				PersistentVolumeClaim: &apiv1.PersistentVolumeClaimVolumeSource{
+					ClaimName: v.Name,
+					ReadOnly:  true,
+				},
+			},
+		}
+		kvs = append(kvs, kv)
+
+		kvm := apiv1.VolumeMount{
+			Name:      v.Name,
+			ReadOnly:  v.ReadOnly,
+			MountPath: v.Mountpoint,
+		}
+		kvms = append(kvms, kvm)
+	}
+
+	_, err = o.Client.CoreV1().Pods(o.Handler.Config.Kubernetes.Namespace).Create(&apiv1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "conplicity-worker",
```

Comment:
Does that name need to be unique? Is there a risk of a naming conflicts if several containers are running at the same time?

Reply:
You're right, I'll generate a random name at the beginning of the function.